lib/kubernetes-deploy/kubernetes_resource/stateful_set.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource/pod_set_base'
+module KubernetesDeploy
+  class StatefulSet < PodSetBase
+    TIMEOUT = 5.minutes
+    ONDELETE = 'OnDelete'
+    attr_reader :pods
+
+    def sync
+      raw_json, _err, st = kubectl.run("get", type, @name, "--output=json")
+      @found = st.success?
+
+      if @found
+        stateful_data = JSON.parse(raw_json)
+        @status_data = stateful_data["status"]
+        @rollout_data = stateful_data["status"].slice("replicas")
+        @update_strategy = stateful_data["spec"]["updateStrategy"]["type"]
+        @status = @rollout_data.map { |state_replicas, num| "#{num} #{state_replicas.chop.pluralize(num)}" }.join(", ")
+        @pods = find_pods(stateful_data)
+      else # reset
+        @rollout_data = { "replicas" => 0 }
+        @status = nil
+        @pods = []
+      end
+    end
+
+    def deploy_succeeded?
+      if @update_strategy == ONDELETE
+        # Gem cannot monitor update since it doesn't occur until delete
+        unless @success_assumption_warning_shown
+          @logger.warn("WARNING: Your SS's updateStrategy is set to OnDelete, "\
+                       "which means updates will not be applied until its deleted. "\
+                       "If you are using k8s 1.7+, consider switching to RollingUpdates.")
+          @success_assumption_warning_shown = true
+        end
+        true
+      else
+        @status_data['current_generation'] == @status_data['observed_generation'] &&
```

Comment:
these field names seem wrong... I'm guessing the test passes because they are both `nil`. 

The real field I see is `.status.observedGeneration`. The current generation is in  `.metadata.generation`, not the status.

Reply:
The fact that SS has so many different status fields is a bit confusing... I went and found an example of a SS that is having trouble rolling out an update to help clarify what we need. The following status belongs to a SS that is in a pretty messed up state:

```yaml
status:
  currentReplicas: 1
  currentRevision: statefulset-3627988389 # there is one pod in this version
  observedGeneration: 4 # this is equal to .metadata.generation... useless?
  readyReplicas: 2 # these are from a revision not mentioned in the status
  replicas: 3
  updateRevision: statefulset-1264314452 # there are no pods in this version
```

And of a healthy SS:

```yaml
status:
  currentReplicas: 2
  currentRevision: other-statefulset-335438845
  observedGeneration: 64
  readyReplicas: 2 
  replicas: 2
  updateRevision: other-statefulset-335438845
```

Based on that, I think we need:
- `.status.currentRevision == .status.updateRevision`
- `.spec.replicas == .status.currentReplicas == .status.readyReplicas`

Reply:
Good catch. The ss fields are confusing in how they are layed out